### PR TITLE
[REFACTOR] [METEOR-1119] Refactor overview / tiled acquisition code

### PR DESCRIPTION
**Pull request by patrickcleeve2**
The previous overview acquisition code was fairly hardcoded for specific fm based imaging and coupled to the gui. The main components are:

**Utilities**:
- Move overview acquisition utils from gui methods into generic utilities allowing them to be used outside the user interface.
- Added test cases for tiling utilities.

**GUI**
- Added support for a acquiring overviews with FIB/SEM, and hide unused settings (e.g. z-stacks) for each mode.
- Allow using AcquireOverview task without autofocus
- Allow multi-area acquisitions without autofocus.
- Consolidate different code paths when using autofocus vs not.
- Removed direct call to AcquireTiledArea task, all paths now go through AcquireOverview.

### CODE REVIEW

File: src/odemis/acq/stitching/_tiledacq.py [hunk 1]
```diff
@@ -62,6 +64,14 @@
 # Increasing this distance may result in out of focus overview acquisition image
 MAX_DISTANCE_FOCUS_POINTS = 450e-06  # in m
 
+# cryo-TEM grid sizes
+SAMPLE_RADIUS_TEM_GRID = 1.25e-3  # m, standard TEM grid size including the borders
+# Bounding-box relative to the center of a sample, corresponding to usable area
+# for imaging/milling. Used in particular for the overview image.
+hwidth = SAMPLE_RADIUS_TEM_GRID / math.sqrt(2)
+SAMPLE_USABLE_BBOX_TEM_GRID = (-hwidth, -hwidth, hwidth, hwidth)  # m, minx, miny, maxx, maxy
+DEFAULT_FOV = (100e-6, 100e-6) # m
```
Comment (K4rishma) [inline src/odemis/acq/stitching/_tiledacq.py:73]:
```suggestion
DEFAULT_FOV = (100e-6, 100e-6)  # m
```

File: src/odemis/acq/stitching/_tiledacq.py [hunk 2]
```diff
@@ -1190,3 +1207,177 @@ def run(self):
                 self._future._task_state = FINISHED
 
         return da_rois
+
+def get_tiled_areas(
```
Comment (K4rishma) [inline src/odemis/acq/stitching/_tiledacq.py:1211]: The new functions like get_tiled_areas, compute area size, the docsting is missing. Do we need test cases for these two functions?

**Review by K4rishma — APPROVED** [review]

File: src/odemis/acq/stitching/_tiledacq.py (hunk 2 above)
Comment (patrickcleeve2) [reply src/odemis/acq/stitching/_tiledacq.py:1211]: added tests and docstrings 👍

File: src/odemis/acq/stitching/_tiledacq.py [hunk 3]
```diff
@@ -1190,3 +1207,198 @@ def run(self):
...
+    w: float, h: float, pos: Dict[str, float], tiling_rng: Dict[str, List[float]]
+) -> Optional[Tuple[float]]:
+    """
+    Finds the intersection between the requested tiling area and the tiling range.
+    :param w: the width of the tiling area
+    :param h: the height of the tiling area
+    :param pos: the current position of the stage
+    :param tiling_rng: the tiling range along x and y axes as
+            (xmin, ymin, xmax, ymax), or (xmin, ymax, xmax, ymin)
+    :return (None or tuple of 4 floats): None if there is no intersection, or
```
Comment (K4rishma) [inline src/odemis/acq/stitching/_tiledacq.py:1337]: ``
```suggestion
    :returns: (None or tuple of 4 floats): None if there is no intersection, or
```

File: src/odemis/acq/stitching/_tiledacq.py [hunk 4]
```diff
@@ -1190,3 +1207,198 @@ def run(self):
...
+
+def clip_tiling_area_to_range(
+    w: float, h: float, pos: Dict[str, float], tiling_rng: Dict[str, List[float]]
+) -> Optional[Tuple[float]]:
+    """
+    Finds the intersection between the requested tiling area and the tiling range.
+    :param w: the width of the tiling area
+    :param h: the height of the tiling area
+    :param pos: the current position of the stage
+    :param tiling_rng: the tiling range along x and y axes as
```
Comment (K4rishma) [inline src/odemis/acq/stitching/_tiledacq.py:1335]: Might be useful to highlight where this tiling range originates from and why do we clip?

**Comment by tepals** [comment]
There is quite a lot moved around, can you write a bit more of a detailed description of what you did? That will make it easier for me to review.

**Comment by patrickcleeve2** [comment]
> There is quite a lot moved around, can you write a bit more of a detailed description of what you did? That will make it easier for me to review.

@tepals I've updated the PR description with more details

File: src/odemis/gui/win/acquisition.py [hunk 5]
```diff
@@ -780,18 +789,36 @@ def __init__(self, parent, orig_tab_data):
         # imprecision, all the tiles will overlap or at worse be next to each other (i.e. , no space between tiles)
         self.overlap = 0.10
         try:
+            if self.fibsem_mode:
+                self.stage = self._main_data_model.stage_bare
+                self.focuser = self._main_data_model.ebeam_focus
+                self.detector = self._main_data_model.sed
+                self.settings_obs = None
```
Comment (tepals) [inline src/odemis/gui/win/acquisition.py:796]: why don't you need a settings observer in fibsem mode?

File: src/odemis/gui/win/acquisition.py [hunk 6]
```diff
@@ -780,18 +789,36 @@ def __init__(self, parent, orig_tab_data):
...
-                "y": self._main_data_model.stage.axes["y"].range
+                "x": self.stage.axes["x"].range,
+                "y": self.stage.axes["y"].range
             }
 
-            stage_md = self._main_data_model.stage.getMetadata()
-            if model.MD_POS_ACTIVE_RANGE in stage_md:
-                self._tiling_rng.update(stage_md[model.MD_POS_ACTIVE_RANGE])
+            stage_md = self.stage.getMetadata()
+            if range_md in stage_md:
```
Comment (tepals) [inline src/odemis/gui/win/acquisition.py:818]: I cannot really think of a better name for `range_md`, but it reads a bit weird that `stage_md` contains all the metadata, while `range_md` is only a single metadata entry. So if you think of a better name, can you update it?

File: src/odemis/acq/stitching/_tiledacq.py [hunk 7]
```diff
@@ -1190,3 +1207,199 @@ def run(self):
...
+            return []
+        else:
+            return [area]
+
+    # TODO: for now this is only for the MIMAS, but eventually, on the METEOR,
+    # which often supports 2 grids, this should also be possible to use.
+    # TODO: ensure that sample centers are converted to same coordinate system as stage position
+
+    # .sample_centers contains the center position of each area
+    # .sample_rel_bbox contains the relative bounding box on an area
```
Comment (tepals) [inline src/odemis/acq/stitching/_tiledacq.py:1257]: do you still have `.sample_centers` and `.sample_rel_bbox` here?

File: src/odemis/acq/stitching/test/tiledacq_test.py [hunk 8]
```diff
@@ -544,6 +545,178 @@ def on_progress_update(self, future, start, end):
...
+        frange = focuser.axes['z'].range
+        max_range = frange[1] - frange[0]
+        zlevels = get_zstack_levels(zsteps=10, zstep_size=max_range / 5, rel=False, focuser=focuser)
+        self.assertAlmostEqual(zlevels[0], frange[0])
+        self.assertAlmostEqual(zlevels[-1], frange[1])
+        self.assertEqual(len(zlevels), 10)
+
+        # shift z levels when outside focus range
+        focuser.moveAbs({"z": frange[0]}).result() # move to the minimum position
+        zlevels = get_zstack_levels(zsteps=10, zstep_size=10e-6 / 5, rel=False, focuser=focuser)
```
Comment (tepals) [inline src/odemis/acq/stitching/test/tiledacq_test.py:718]: why do you divide the zstep_size by 5?

**Review by tepals — CHANGES_REQUESTED** [review]

File: src/odemis/gui/win/acquisition.py (hunk 6 above)
Comment (patrickcleeve2) [reply src/odemis/gui/win/acquisition.py:818]: changed to imaging_range, as its an an artificical limit on the stage for imaging

**Comment by patrickcleeve2** [comment]
@nandishjpatel I've added you as reviewer, because @tepals is on leave for a few weeks

File: src/odemis/acq/stitching/_tiledacq.py [hunk 9]
```diff
@@ -1036,6 +1044,7 @@ def __init__(self, streams, stage, areas, focus, ccd, future=None, overlap=0.2,
 
         self.conf_level = 0.8
         self.focusing_method = focusing_method
+        self._use_autofocus: bool = use_autofocus
```
Comment (K4rishma) [inline src/odemis/acq/stitching/_tiledacq.py:1047]: Is it useful to explicitly mention it as bool as use_autofocus is defined as bool already?

**Review by K4rishma — APPROVED** [review]

File: src/odemis/acq/stitching/_tiledacq.py [hunk 10]
```diff
@@ -1108,37 +1119,40 @@ def run(self):
...
+                                                                    self.focus_rng,
+                                                                    self._focus_points[idx],
+                                                                    self.conf_level)
+
+                    try:
+                        focus_points_per_area = len(self._focus_points[idx])
+                        max_wait_t = estimate_autofocus_in_roi_time(focus_points_per_area, self._det) * 3 + 1
+                        focus_points = self._future.running_subf.result(max_wait_t)
+                    except TimeoutError:
+                        self.streams[0].is_active.value = False
```
Comment (nandishjpatel) [inline src/odemis/acq/stitching/_tiledacq.py:1149]:
```suggestion
                        self.streams[0].is_active.value = False
```
Can this go in `finally`?

File: src/odemis/gui/win/acquisition.py [hunk 11]
```diff
@@ -1158,43 +1128,7 @@ def _get_zstack_levels(self, rel: bool = False):
...
-
-        # TODO: if there is an even number of zsteps, the current focus position will not be part of the zlevels.
-        # As the current focus position can often be the "best" position, we could change the algorithm in such case to
-        # shift exactly symmetrical, but instead use the current position as one of the zlevels.
-
-        # Create focus zlevels from the given zsteps number
-        zlevels = numpy.linspace(zmin, zmax, zsteps).tolist()
-
-        return zlevels
+        return get_zstack_levels(zsteps=self.zsteps.value, zstep_size=self.zstep_size.value, rel=rel, focuser=self.focuser)
```
Comment (nandishjpatel) [inline src/odemis/gui/win/acquisition.py:1131]: Why not directly use `get_zstack_levels` instead of `_get_zstack_levels`?

Comment (nandishjpatel) [reply src/odemis/gui/win/acquisition.py:1131]: Ahh to avoid the args I guess

File: src/odemis/acq/stitching/_tiledacq.py [hunk 12]
```diff
@@ -1190,3 +1207,196 @@ def run(self):
...
+            tiles_ny=tiles_ny,
+            tiling_rng=tiling_rng,
+            overlap=overlap,
+        )
+        if area is None:
+            return []
+        else:
+            return [area]
+
+    # TODO: for now this is only for the MIMAS, but eventually, on the METEOR,
```
Comment (nandishjpatel) [inline src/odemis/acq/stitching/_tiledacq.py:1252]: Is this still a `TODO`?

File: src/odemis/acq/stitching/_tiledacq.py [hunk 13]
```diff
@@ -1190,3 +1207,196 @@ def run(self):
...
+    :return: the size (in m) in X and Y. If no area at all, returns None.
+    """
+    # get smallest fov
+    fovs = [get_fov(s) for s in streams]
+    if not fovs:
+        # fall back to a small fov (default)
+        fov = DEFAULT_FOV
+    else:
+        # smallest fov
+        fov = (min(f[0] for f in fovs), min(f[1] for f in fovs))
```
Comment (nandishjpatel) [inline src/odemis/acq/stitching/_tiledacq.py:1300]:
```suggestion
        fov = tuple(map(min, zip(*fovs)))
```

**Review by nandishjpatel — APPROVED** [review]
Some minor comments, else looking good

File: src/odemis/acq/stitching/_tiledacq.py (hunk 12 above)
Comment (patrickcleeve2) [reply src/odemis/acq/stitching/_tiledacq.py:1252]: Yes, the sample centers still aren't implemented correctly for meteor

File: src/odemis/acq/stitching/_tiledacq.py (hunk 10 above)
Comment (patrickcleeve2) [reply src/odemis/acq/stitching/_tiledacq.py:1149]: good suggestion

**Review by tepals — APPROVED** [review]

**Pull request merged**